### PR TITLE
Implement new architecture, modifier stacks

### PR DESCRIPTION
**Pull request by kducharm**
https://www.drupal.org/project/migration_tools/issues/2953486#comment-12543558

- Creates source (HTML) operations and DOM (QueryPath) operations, which allow for ordering of operations mixed with field retrieval from the DOM. 

- Adds a DOM migrate_plus data_parser_plugin, which allow you to extract (chunk) URLs from an HTML page into rows for migration. 

- Adds ability to use jobs to set a variable, then use that variable as a dynamic argument in another job's arguments

- Adds replaceString modifier.

- Adds HREF functions ObtainLink/ObtainFileLink

- Updates examples

### CODE REVIEW

File: migration_tools_example/config/install/migrate_plus.migration_tools_example_drupal_import.yml [hunk 1]
```diff
@@ -5,50 +5,74 @@ migration_group: migration_tools_example
...
+        -
+          operation: modifier
+          modifier: removeSelectorAll
+          arguments:
+            - '.pagedate'
+            - 'head'
+        -
+          # Get the body field after above modifiers have run.
+          operation: get_field
+          field: body
```
Comment (swirtSJW) [inline migration_tools_example/config/install/migrate_plus.migration_tools_example_drupal_import.yml]: I really like how readable that was.    It was clear to see what was going on.  Nice example

File: src/Obtainer/ObtainArray.php [hunk 2]
```diff
@@ -26,7 +27,18 @@ public static function cleanString($found) {
     $found = (empty($found)) ? [] : $found;
     // Make sure it is an array, just in case someone uses a string finder.
     $found = (is_array($found)) ? $found : [$found];
-    $found = array_map('\Drupal\migration_tools\StringTools::superTrim', $found);
+
+    // Only run if the value is not an array.
+    $found = array_map(
```
Comment (swirtSJW) [inline src/Obtainer/ObtainArray.php:32]: NIce use of closure.

**Comment by swirtSJW** [comment]
I really like how this is shaping up.   It is much cleaner than before and I think leaves more options open for more use cases.   Well done.

**Comment by kducharm** [comment]
@swirtSJW Can you review the last 3 commits, the examples weren't going to work with drupal.org since they block bots/scripts - and I found a bug w/the DOM parser that prevented it from working with the examples. They are working now with enabling the migration_tools_example module.

This is ready to merge once approved, and would like to do another alpha release linked to the "stacked modifiers" issue

**Pull request merged**